ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -130,6 +130,58 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
+		}
+		// If there is no cluster instance and no error, suspend the job deployment
+		if rayClusterInstance == nil {
+			// TODO (kevin85421): Note that if we suspend the RayJob in the `Initializing` status, the RayJob will not
+			// transition to the `Suspending` status. It will transition to the `Suspended` status directly. We should
+			// unify the code path.
+			err = r.updateState(ctx, rayJobInstance, nil, rayJobInstance.Status.JobStatus, rayv1.JobDeploymentStatusSuspended)
+			if err != nil {
+				return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
+			}
+			r.Log.Info("RayJob suspended", "RayJob", rayJobInstance.Name)
+			r.Recorder.Eventf(rayJobInstance, corev1.EventTypeNormal, "Suspended", "Suspended RayJob %s", rayJobInstance.Name)
+			return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
+		}
+
+		// Check the current status of RayCluster before submitting.
+		if clientURL := rayJobInstance.Status.DashboardURL; clientURL == "" {
+			if rayClusterInstance.Status.State != rayv1.Ready {
+				r.Log.Info("Wait for the RayCluster.Status.State to be ready before submitting the job.", "RayCluster", rayClusterInstance.Name, "State", rayClusterInstance.Status.State)
+				return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
+			}
+
+			if clientURL, err = utils.FetchHeadServiceURL(ctx, &r.Log, r.Client, rayClusterInstance, utils.DashboardPortName); err != nil || clientURL == "" {
+				r.Log.Error(err, "Failed to get the dashboard URL after the RayCluster is ready!", "RayCluster", rayClusterInstance.Name)
+				return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
+			}
+			rayJobInstance.Status.DashboardURL = clientURL
+
+			// TODO (kevin85421): The function `updateState` might skip the update if `JobStatus` and `JobDeploymentStatus`
+			// are already identical to the given values. Therefore, we use `r.Status().Update()` to directly update the status.
+			// However, neither `updateState` nor `r.Status().Update()` is an ideal solution. We should refactor the code.
+			if err := r.Status().Update(ctx, rayJobInstance); err != nil {
+				r.Log.Error(err, "Failed to update the dashboard URL to the RayJob status!", "RayJob", rayJobInstance.Name)
+			}
+			return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
+		}
+
+		// Ensure k8s job has been created
+		if err := r.createK8sJobIfNeed(ctx, rayJobInstance, rayClusterInstance); err != nil {
+			return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
+		}
+
+		r.Log.Info("Both RayCluster and the submitter K8s Job are created. Transition the status from `Initializing` to `Running`.",
+			"RayJob", rayJobInstance.Name, "RayCluster", rayJobInstance.Status.RayClusterName)
+		if err = r.updateState(ctx, rayJobInstance, nil, rayJobInstance.Status.JobStatus, rayv1.JobDeploymentStatusRunning); err != nil {
+			return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
```

Comment:
in this case what happens if reconciler runs after `RayJobDefaultRequeueDuration`. Would we create duplicate cluster in line #135?

Reply:
There is a queue that stores resource events capable of triggering reconciliation. `RequeueAfter: RayJobDefaultRequeueDuration` implies that an event for this RayJob custom resource is placed in the queue to trigger reconciliation after `RayJobDefaultRequeueDuration` seconds.

> Would we create duplicate cluster in line 135 ?

There is a delay between our local cache and the Kubernetes API server. This means the RayCluster may have already been created in the Kubernetes API server, but our local cache hasn't completed synchronization yet. Therefore, we call the Create function again to create the RayCluster.

To avoid the RayCluster double creation, we use the same RayCluster name for the same RayJob custom resource for each reconciliation loop. Hence, if we call the `Create` function twice, the second one will be denied by the Kubernetes API server because the RayCluster already exists.

Reply:
aah cool. Thanks for confirming that